models/workitem_repository.go

```diff
@@ -1,214 +1,211 @@
+	}
+
+	log.Printf("loading work item %d", id)
+	res := WorkItem{}
+	if r.ts.tx.First(&res, id).RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, NotFoundError{"work item", ID}
+	}
+	wiType, err := r.wir.loadTypeFromDB(ctx, res.Type)
+	if err != nil {
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+	result, err := convertFromModel(*wiType, res)
+	if err != nil {
+		return nil, ConversionError{simpleError{err.Error()}}
+	}
+	return result, nil
+}
+
+// Delete deletes the work item with the given id
+// returns NotFoundError or InternalError
+func (r *GormWorkItemRepository) Delete(ctx context.Context, ID string) error {
+	var workItem = WorkItem{}
+	id, err := strconv.ParseUint(ID, 10, 64)
+	if err != nil {
+		// treat as not found: clients don't know it must be a number
+		return NotFoundError{entity: "work item", ID: ID}
+	}
+	workItem.ID = id
+	tx := r.ts.tx
+
+	if err = tx.Delete(workItem).Error; err != nil {
+		if tx.RecordNotFound() {
+			return NotFoundError{entity: "work item", ID: ID}
+		}
+		return InternalError{simpleError{err.Error()}}
+	}
+
+	return nil
+}
+
+// Save updates the given work item in storage. Version must be the same as the one int the stored version
+// returns NotFoundError, VersionConflictError, ConversionError or InternalError
+func (r *GormWorkItemRepository) Save(ctx context.Context, wi app.WorkItem) (*app.WorkItem, error) {
+	res := WorkItem{}
+	id, err := strconv.ParseUint(wi.ID, 10, 64)
+	if err != nil {
+		return nil, NotFoundError{entity: "work item", ID: wi.ID}
+	}
+
+	log.Printf("looking for id %d", id)
+	tx := r.ts.tx
+	if tx.First(&res, id).RecordNotFound() {
+		log.Printf("not found, res=%v", res)
+		return nil, NotFoundError{entity: "work item", ID: wi.ID}
+	}
+	if res.Version != wi.Version {
+		return nil, VersionConflictError{simpleError{"version conflict"}}
+	}
+
+	wiType, err := r.wir.loadTypeFromDB(ctx, wi.Type)
+	if err != nil {
+		return nil, BadParameterError{"Type", wi.Type}
+	}
+
+	newWi := WorkItem{
+		ID:      id,
+		Type:    wi.Type,
+		Version: wi.Version + 1,
+		Fields:  Fields{},
+	}
+
+	for fieldName, fieldDef := range wiType.Fields {
+		fieldValue := wi.Fields[fieldName]
+		var err error
+		newWi.Fields[fieldName], err = fieldDef.ConvertToModel(fieldName, fieldValue)
+		if err != nil {
+			return nil, BadParameterError{fieldName, fieldValue}
+		}
+	}
+
+	if err := tx.Save(&newWi).Error; err != nil {
+		log.Print(err.Error())
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+	log.Printf("updated item to %v\n", newWi)
+	result, err := convertFromModel(*wiType, newWi)
+	if err != nil {
+		return nil, InternalError{simpleError{err.Error()}}
+	}
+	return result, nil
+}
+
+// Create creates a new work item in the repository
+// returns BadParameterError, ConversionError or InternalError
+func (r *GormWorkItemRepository) Create(ctx context.Context, typeID string, fields map[string]interface{}) (*app.WorkItem, error) {
```

Comment:
@aslakknutsen @tsmaeder Again, can't we use `fields map[string]FieldDefinition` here?